RoT/Protocol/RFCs/0001-Transport_Agnosticism.md

```diff
@@ -0,0 +1,149 @@
+   - Addressing information that determines which device sent the message
+     (the actual contents of this information is irrelevant to Cerberus).
+
+4. Messages are "addressed", in two respects:
+   a. A server can take a request and reply to the original sender with a
+      response, and a client can match up requests and their responses.
+   b. Cerberus can use addressing information of unspecified format to
+      construct requests to a specific device (this opaque addressing
+      information could be present in a PCD, for example).
+
+By construction, MCTP almost fulills the above requirements: it provides framing
+of a message of arbitrary size, and the PCIe Vendor command provides a location
+for the transport-specific header. Note that Cerberus itself performs the
+size negotiation at the moment, so that would be moved out of the Challenge
+Protocol and into the transport layer.
+
+The above list is everything Cerberus needs: Cerberus can be spoken over any
+admissible transport with not change to the actual bytes present in the
+encoded messages.
+
+# Specification Changelist
+
+This RFC seeks the following changes to the Challenge Protocol spec:
+
+- Chapter 3 (Protocol and Hierarchy) should be replaced with the list of
+  requirements given above. We do not recommend using the above text
+  exactly; instead, the new Chapter 3 should carefully elaborate each point,
+  to make it possible for third parties to evaluate whether their chosen
+  transport layer is admissible.
+
+- The current contents of Chapter 3 should be moved into an appendix, provided
+  as an example admissible transport layer for I2C connections. The same should
+  be done for Section 8.1, which describes a legacy protocol built on top of
+  SMBus.
+
+- The first few sections of Chapter 6 should be modified to not reference MTCP,
+  and instead refer to the abstract Cerberus header (request bit, type, length)
+  derived from the transport-specific header.
+
+- The MCTP-specific errors in Section 6.5 should be either removed, or made
+  transport agnostic. For example, the packet/message length overflow errors
+  should be handled at the transport protocol layer: in the case that a device
+  sends a too-large packet, Cerberus should not be notified at all.
+
+- Mentions of MCTP in Section 6.7 should be removed. The maximum message/packet
+  size fields should be removed, since the transport layer should negotiate
+  these while establishing a session, and pass the maximum message size field
+  along to Cerberus.
+
+- The same should be done for Section 6.13.
+
+- The same should also be done for Section 6.18. Since the response makes
+  reference to payload sizes, we recommend that, instead of replying with
+  multiple messages of maximum size, we add a "continue" bit to the Get Log
+  response. When this bit is set, the client must send another request, with
+  the offset field increased by the length of the Get Log log contents field.
```

Comment:
I think we can leave the same message structure and semantics while achieving the goal of removing message sizes from the Challenge Protocol layer.  It could be reworded to something like this:

"The amount of log information returned is determined by the actual log size or the maximum message size as determined by the protocol transport, whichever is smaller.  If multiple messages are required to retrieve the full log contents, the end of log data is denoted by a response message with less than the transport-defined maximum number of bytes.  A request for log data with the Offset field set to equal or greater than the log size will generate a response message of 0 bytes."

What do you think?

Reply:
I don't *love* this, but backwards compatibility in binary format is attractive... I want to sleep on it.